### DISCUSSION TITLE
Cancelling order if not filled after bar duration

### DISCUSSION THREAD

**Discussion by julzKB** [kernc/backtesting.py#567] [Q&A] [answered]
Hello there,

As per the documentation, orders are placed GTC and need to be cancelled manually if not filled . What would be the best way to do so if the order is not filled after the duration of one bar ?

Thank you

**Comment by kernc** [comment, marked as answer]
Best to look in the [API reference](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html). There is [`Strategy.orders`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.orders) and [`Order.cancel`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order.cancel), so something like this should work:
```py
    def next(self):
        # Close previous bar's orders *before* opening any new ones
        for order in self.orders:
            order.cancel()

        # Open new orders for this bar
        ...
```

**Reply by julzKB** [reply]
That makes sense, I guess I'm confused between when I issue the buy/stop signal and making sure the duration of a bar passes before I cancel it .

In the example below I issue a buy order but it never has time to become a trade since I think I am canceling the order not leaving enough time for the trade to kick in if that makes sense :

```
    def init(self):
        pattern = ta.CDLHARAMI
        close = self.data.Close
        Open = self.data.Open
        high = self.data.High
        low = self.data.Low
        self.pattern = self.I(pattern, Open, high, low, close)

        

    def next(self):
        if self.pattern == 100:
            stop = self.data.High[-1]
            sl = self.data.Low[-1]
            tp = self.data.High[-1]+(self.data.High[-1]-self.data.Low[-1])
            self.buy(stop=stop, sl=sl,tp=tp)

        for order in self.orders:
            order.cancel()
```

**Reply by kernc** [reply]
Please see my revised answer above.

**Reply by julzKB** [reply]
haha, yes, that makes sense now, thank you :)
